### PR TITLE
Adding categorized block and raw block structs

### PR DESCRIPTION
**Pull request by cloudnoize**
Block contains data per category, about keys that were added at this block.
In addition, it contains the parent digest and the block ID.

Each category is responsible to handle the block struct its ```update_info``` as the result of an update operation.

Rawblock is a unit, which we can transfer and reconstruct a block in a remote replica.
It contains all the relevant updates as received from the client in the initial call to storage.

RawBlock can be constructed from a ```Block``` struct.

In addition, this PR adds support to ```unsigned char``` in RocksDB NativeClient.
Refactoring to the updates and kv_blockchain API.
Unit tests.

### CODE REVIEW

File: kvbc/cmf/categorized_kvbc_msgs.cmf [hunk 1]
```diff
@@ -22,39 +22,88 @@ Msg SharedValueData 4{
...
+Msg SharedKeyValueUpdatesData 5{
     map string SharedValueData kv
     bool calculate_hash
 }
 
 
 # Updates info
 
-Msg MerkleKeyFlag 6{
+Msg MerkleKeyFlag 10{
```
Comment (dartdart26) [inline kvbc/cmf/categorized_kvbc_msgs.cmf:33]: Are the ID gaps used to allow new messages of different types? If yes, maybe we can use a much bigger gaps, something like 100 or 1000. What do you think?

File: kvbc/cmf/categorized_kvbc_msgs.cmf [hunk 2]
```diff
@@ -22,39 +22,88 @@ Msg SharedValueData 4{
...
     bool deleted
 }
 
-Msg MerkleUpdatesInfo 7{
+Msg MerkleUpdatesInfo 11{
     map string MerkleKeyFlag keys
-    fixedlist uint8 32 hash
-    uint64 version
+    fixedlist uint8 32 state_hash
+    uint64 state_root_version
```
Comment (dartdart26) [inline kvbc/cmf/categorized_kvbc_msgs.cmf:40]: Nit: Maybe these can be unified in a new type as they go hand in hand.

File: kvbc/cmf/categorized_kvbc_msgs.cmf [hunk 3]
```diff
@@ -22,39 +22,88 @@ Msg SharedValueData 4{
...
+    optional SharedKeyValueUpdatesInfo shared_updates_info
+    map string oneof{
+        MerkleUpdatesInfo
+        KeyValueUpdatesInfo
+    } categories_updates_info
+}
+
+Msg RawBlockMerkleUpdates 22 {
+    MerkleUpdatesData updates
+    fixedlist uint8 32 state_hash
```
Comment (dartdart26) [inline kvbc/cmf/categorized_kvbc_msgs.cmf:80]: Nit: Maybe we can call all hashes for all categories `root_hash`.

File: kvbc/include/categorization/blocks.h [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+
+// Block is composed out of:
+// - BlockID
+// - Parent digest
+// - Categories' updates_info where update info is essentially a list of keys, metadata on the keys and a hash of the
+// category state.
+struct Block {
+  static const std::string CATEGORY_ID;
+
+  Block() = default;
```
Comment (dartdart26) [inline kvbc/include/categorization/blocks.h:19]: Nit: Maybe initialise the block ID to 0 in that case?

File: kvbc/include/categorization/blocks.h [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+  Block(const BlockId block_id) { data.block_id = block_id; }
+
+  void add(const std::string& category_id, MerkleUpdatesInfo&& updates_info) {
+    data.categories_updates_info.emplace(std::make_pair(category_id, std::move(updates_info)));
```
Comment (dartdart26) [inline kvbc/include/categorization/blocks.h:23]: Nit: I think we can omit `std::make_pair()` as we are doing emplace() and we can just pass the arguments.

File: kvbc/include/categorization/blocks.h [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+  void add(const std::string& category_id, KeyValueUpdatesInfo&& updates_info) {
+    data.categories_updates_info.emplace(std::make_pair(category_id, std::move(updates_info)));
+  }
+
+  void add(SharedKeyValueUpdatesInfo&& updates_info) { data.shared_updates_info = std::move(updates_info); }
+
+  void setParentHash(const BlockDigest& parent_hash) {
+    std::copy(parent_hash.begin(), parent_hash.end(), data.parent_digest.begin());
+  }
+  inline BlockId ID() const { return data.block_id; }
```
Comment (dartdart26) [inline kvbc/include/categorization/blocks.h:35]: Nit: Naming, think should be `id()`.

File: kvbc/include/categorization/blocks.h [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+
+  static const serialization::buffer& serialize(const Block& block) {
+    static thread_local serialization::buffer output;
```
Comment (dartdart26) [inline kvbc/include/categorization/blocks.h:38]: Nice! :)

File: kvbc/include/categorization/blocks.h [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+  }
+
+  void add(SharedKeyValueUpdatesInfo&& updates_info) { data.shared_updates_info = std::move(updates_info); }
+
+  void setParentHash(const BlockDigest& parent_hash) {
+    std::copy(parent_hash.begin(), parent_hash.end(), data.parent_digest.begin());
+  }
+  inline BlockId ID() const { return data.block_id; }
+
+  static const serialization::buffer& serialize(const Block& block) {
```
Comment (dartdart26) [inline kvbc/include/categorization/blocks.h:37]: Btw, can that be a member (non-static) method?

File: kvbc/include/categorization/blocks.h [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+//////////////////////////////////// RAW BLOCKS//////////////////////////////////////
+
+// Reconstructs the updates data as recieved from the user
+// This set methods are overloaded in order to construct the appropriate updates
+
+// Merkle updates reconstruction
+RawBlockMerkleUpdates getRawUpdates(const std::string& category_id,
+                                    const MerkleUpdatesInfo& update_info,
+                                    const BlockId& block_id,
+                                    std::shared_ptr<storage::rocksdb::NativeClient>& native_client) {
```
Comment (dartdart26) [inline kvbc/include/categorization/blocks.h:75]: Nit: As there's no ownership and no write ops, we can just pass `const NativeClient*`.

File: kvbc/include/categorization/blocks.h [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+// shared updates reconstruction
+RawBlockSharedUpdates getRawSharedUpdates(const SharedKeyValueUpdatesInfo& update_info,
+                                          const BlockId& block_id,
+                                          std::shared_ptr<storage::rocksdb::NativeClient>& native_client) {
+  RawBlockSharedUpdates data;
+  return data;
+}
+
+// Raw block is the unit which we can transfer and reconstruct a block in a remote replica.
+// It contains all the relevant updates as recieved from the client in the initial call to storage.
```
Comment (dartdart26) [inline kvbc/include/categorization/blocks.h:126]: Nice comment!

File: kvbc/include/categorization/blocks.h [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+// It also contains:
+// - parent digest
+// - state hash per category (if exists) E.L check why do we pass it.
+struct RawBlock {
+  RawBlock(const Block& block, std::shared_ptr<storage::rocksdb::NativeClient> native_client) {
```
Comment (dartdart26) [inline kvbc/include/categorization/blocks.h:131]: Nit: Maybe we don't need to bump the ref count on the native client (i.e. use non-owning pointer, as mentioned above).

File: kvbc/include/categorization/blocks.h [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+    for (auto& [cat_id, update_info] : block.data.categories_updates_info) {
+      std::visit(
+          [category_id = cat_id, &block, this, &native_client](const auto& update_info) {
+            auto category_updates = getRawUpdates(category_id, update_info, block.ID(), native_client);
+            data.category_updates.emplace(std::make_pair(category_id, std::move(category_updates)));
+          },
+          update_info);
+    }
+    if (block.data.shared_updates_info.has_value()) {
+      auto category_updates = getRawSharedUpdates(block.data.shared_updates_info.value(), block.ID(), native_client);
```
Comment (dartdart26) [inline kvbc/include/categorization/blocks.h:144]: Nit: Maybe no need for the `category_updates` variable.

File: kvbc/include/categorization/blocks.h [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+struct RawBlock {
+  RawBlock(const Block& block, std::shared_ptr<storage::rocksdb::NativeClient> native_client) {
+    // parent digest (std::copy?)
+    data.parent_digest = block.data.parent_digest;
+    // recontruct updates of categories
+    for (auto& [cat_id, update_info] : block.data.categories_updates_info) {
+      std::visit(
+          [category_id = cat_id, &block, this, &native_client](const auto& update_info) {
+            auto category_updates = getRawUpdates(category_id, update_info, block.ID(), native_client);
+            data.category_updates.emplace(std::make_pair(category_id, std::move(category_updates)));
```
Comment (dartdart26) [inline kvbc/include/categorization/blocks.h:139]: Nit: Maybe no need for `std::make_pair()` as mentioned above.

File: kvbc/include/categorization/details.h [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+
+namespace concord::kvbc::categorization {
+
+namespace serialization {
+using buffer = std::vector<uint8_t>;
+}  // namespace serialization
+
+// DB key is a CMF stuct of hashed key and the version.
+// On serialization it will be hashed_key || version(big endian)
+DbKey genDbKey(const std::string& key, const BlockId& version) {
```
Comment (dartdart26) [inline kvbc/include/categorization/details.h:15]: Is that function related to the merkle category? If not, maybe we can be explicit about which hash it uses (i.e. `concord::util::SHA3_256`. Also, maybe we can define a hash type to reuse across `categorization`.

File: storage/include/rocksdb/native_client.h [hunk 15]
```diff
@@ -237,6 +237,17 @@ class NativeClient : public std::enable_shared_from_this<NativeClient> {
...
   }
 
+  template <
+      typename Span,
+      std::enable_if_t<std::is_convertible_v<decltype(std::declval<Span>().size()), std::size_t> &&
+                           std::is_pointer_v<decltype(std::declval<Span>().data())> &&
+                           std::is_convertible_v<std::remove_pointer_t<decltype(std::declval<Span>().data())> (*)[],
+                                                 const uint8_t (*)[]>,
+                       int> = 0>
+  static ::rocksdb::Slice toSlice(const Span &span) noexcept {
```
Comment (dartdart26) [inline storage/include/rocksdb/native_client.h:247]: Nice!

Nit: Could we please update the comment at the top of the file that mentions NativeClient only accepts vector<char>, span<char> and so on to vector<char/uint8_t>, span<char/uint8_t>.

File: kvbc/include/categorization/updates.h [hunk 16]
```diff
@@ -143,26 +143,27 @@ struct MerkleUpdates {
...
                              category_id};
     }
   }
 
-  void addKeyValueUpdates(std::string category_id, KeyValueUpdates&& updates) {
-    if (const auto [itr, inserted] = category_updates_.insert({category_id, std::move(updates.data_)}); inserted) {
+  void add(std::string&& category_id, KeyValueUpdates&& updates) {
+    if (const auto [itr, inserted] = category_updates_.try_emplace(std::move(category_id), std::move(updates.data_));
+        !inserted) {
       *itr;  // disable unused variable
```
Comment (dartdart26) [inline kvbc/include/categorization/updates.h:158]: Ah, I wish one could use `[[maybe_unused]]` with structured bindings. Maybe we can use the `(void)itr` hack to avoid any possible runtime cost from `*`. Not that the compiler will do something on that line in O3...

File: kvbc/include/categorization/updates.h [hunk 17]
```diff
@@ -171,10 +172,12 @@ struct Updates {
     return category_updates_;
   }
 
-  std::optional<SharedUpdatesData>& getSharedUpdates() { return shared_update_; }
+  std::optional<SharedKeyValueUpdatesData>& getSharedUpdates() { return shared_update_; }
```
Comment (dartdart26) [inline kvbc/include/categorization/updates.h:175]: I think we should remove that method to avoid users calling it (now that `KeyValueBlockchain` is a friend).

File: kvbc/include/categorization/kv_blockchain.h [hunk 18]
```diff
@@ -16,64 +16,87 @@
...
 namespace concord::kvbc::categorization {
 
-struct Block {
-  void add(const std::string& category_id, std::variant<MerkleUpdatesInfo, KeyValueUpdatesInfo> updates_info) {}
-  void add(SharedKeyValueUpdatesInfo updates_info) {}
-};
 class KeyValueBlockchain {
  public:
   KeyValueBlockchain() : native_client_(NativeClient::newClient("/tmp", false, NativeClient::DefaultOptions{})) {}
+  KeyValueBlockchain(std::shared_ptr<NativeClient> native_client) : native_client_(native_client) {}
```
Comment (dartdart26) [inline kvbc/include/categorization/kv_blockchain.h:30]: Nit: `const std::shared_ptr<>&` ?

File: kvbc/include/categorization/kv_blockchain.h [hunk 19]
```diff
@@ -16,64 +16,87 @@
...
   // 1) Defines a new block
   // 2) calls per cateogry with its updates
   // 3) inserts the updates KV to the DB updates set per column family
   // 4) add the category block data into the new block
-  BlockId addBlock(Updates& updates) {
+  BlockId addBlock(Updates&& updates) {
```
Comment (dartdart26) [inline kvbc/include/categorization/kv_blockchain.h:35]: Wow, that function got simpler now :)

File: kvbc/test/categorization/updates_test.cpp [hunk 20]
```diff
@@ -23,10 +24,131 @@ TEST(categorized_updates, merkle_update) {
   ASSERT_TRUE(mu.getData().kv.size() == 1);
 }
 
-TEST(categorized_updates, add_block) {
-  KeyValueBlockchain block_chain;
-  Updates updates;
-  ASSERT_TRUE(block_chain.addBlock(updates) == (BlockId)8);
+TEST(categorized_updates, add_blocks) {
+  auto native_client = NativeClient::newClient("/tmp/add_block", false, NativeClient::DefaultOptions{});
```
Comment (dartdart26) [inline kvbc/test/categorization/updates_test.cpp:28]: The storage_test_common.h file contains native client handling and cleanup. Can be reused here too.

**Review by dartdart26 — APPROVED** [review]
Thanks @cloudnoize ! Looks great! I have put comments that you can decide whether to handle or not or do it later.

File: kvbc/cmf/categorized_kvbc_msgs.cmf (hunk 1 above)
Comment (cloudnoize) [reply kvbc/cmf/categorized_kvbc_msgs.cmf:33]: I didn't see it complain so I guess it's allowed. I'll enlarge the gap

File: kvbc/cmf/categorized_kvbc_msgs.cmf (hunk 3 above)
Comment (cloudnoize) [reply kvbc/cmf/categorized_kvbc_msgs.cmf:80]: yep

File: kvbc/include/categorization/blocks.h (hunk 4 above)
Comment (cloudnoize) [reply kvbc/include/categorization/blocks.h:19]: true

File: kvbc/include/categorization/blocks.h (hunk 8 above)
Comment (cloudnoize) [reply kvbc/include/categorization/blocks.h:37]: yeah, I made it on purpose, seems nice to have as a utility function under the Block name.
what do you think?

File: kvbc/include/categorization/blocks.h (hunk 9 above)
Comment (cloudnoize) [reply kvbc/include/categorization/blocks.h:75]: yep

File: kvbc/include/categorization/blocks.h (hunk 11 above)
Comment (cloudnoize) [reply kvbc/include/categorization/blocks.h:131]: yep, I used to work like this, but some people didn't like it :)

File: kvbc/include/categorization/details.h (hunk 14 above)
Comment (cloudnoize) [reply kvbc/include/categorization/details.h:15]: it's not related to the Merkle, it's related to all but the Merkle :) as they can use the new CMF.
So yes we can define a category hash type and use it.

File: kvbc/include/categorization/blocks.h (hunk 11 above)
Comment (dartdart26) [reply kvbc/include/categorization/blocks.h:131]: Bad people :) I mean we just use the client in the scope, we don't save it as a member or anything. I think that's the typical pattern for C++. If we were saving it somewhere that's a different topic, but this ctor is just a function.

Better yet - we can use a reference! :)

File: kvbc/include/categorization/kv_blockchain.h (hunk 18 above)
Comment (cloudnoize) [reply kvbc/include/categorization/kv_blockchain.h:30]: indeed

File: kvbc/include/categorization/blocks.h (hunk 8 above)
Comment (dartdart26) [reply kvbc/include/categorization/blocks.h:37]: I was thinking you need an object to call it and maybe you can make it a `const` method. No big deal.
